### PR TITLE
Implement CosineSimilarity

### PR DESCRIPTION
**Pull request by mejai1206**
## 🙏 Describe the pull request

CosineSimilarity 를 구현했습니다.

## 💬 Additional context
측정 기기: a30

순전파
순전파

역전파
역전파

## ✅ Checklist

- [x] Code follows the project's coding conventions and style.
- [x] Tests have been added or updated to cover the changes.
- [x] Documentation has been updated, if necessary.

### CODE REVIEW

File: benchmarks/benchmark_cosine_similarity.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+@util.report(
+    "cosine_similarity forward",
+    ["x_size"],
+    [256 * i for i in range(1, 21)],
+    {"z_size": 16, "y_size": 16},
+)
+def bench_cosine_similarity_forward(z_size, y_size, x_size, ctx):
+    factory_kwargs = {"device": "cuda"}
+
+    input1 = torch.randn(z_size, y_size, x_size, **factory_kwargs)
```
Comment (daemyung) [inline benchmarks/benchmark_cosine_similarity.py]: 변수명 `input`, `other`가 더 적절해 보입니다.

File: benchmarks/benchmark_cosine_similarity.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            lambda: torch.nn.functional.cosine_similarity(input1, input2, 2)
+        )
+    else:
+        return triton.testing.do_bench(
+            lambda: trident.function.consine_similarity(input1, input2, 2)
+        )
+
+
+@util.report(
+    "cosine_similarity backward",
```
Comment (daemyung) [inline benchmarks/benchmark_cosine_similarity.py]: 타이틀에는 `_`을 사용하지 않습니다. `cosine_similarity`로 변경해 주세요.

File: benchmarks/benchmark_cosine_similarity.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+import torch
+import triton
+import util
+
+import trident
+
+
+@util.report(
+    "cosine_similarity forward",
```
Comment (daemyung) [inline benchmarks/benchmark_cosine_similarity.py]: 타이틀에는 `_`을 사용하지 않습니다. `cosine_similarity`로 변경해 주세요.

File: benchmarks/benchmark_cosine_similarity.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    ["x_size"],
+    [256 * i for i in range(1, 21)],
+    {"z_size": 16, "y_size": 16},
```
Comment (daemyung) [inline benchmarks/benchmark_cosine_similarity.py]: `z_size`는 배치인가요?

File: tests/test_cosine_similarity.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+
+@pytest.mark.parametrize(
+    "z_size, y_size, x_size, dim",
+    [(5431, 500, 200, 0), (221, 1250, 200, 1), (21, 6400, 86, 2)],
+)
+def test_forward(z_size, y_size, x_size, dim, device, dtype):
+    factory_kwargs = {"device": device, "dtype": dtype}
+
+    input1 = torch.randn(z_size, y_size, x_size, **factory_kwargs)
```
Comment (daemyung) [inline tests/test_cosine_similarity.py]: 변수명 `input`, `other`가 더 적절해 보입니다.

File: trident/function/function.py [hunk 6]
```diff
@@ -47,6 +47,10 @@ def conv2d(input, weight, bias=None):
     return operation.Conv2d.apply(input, weight, bias)
 
 
+def consine_similarity(input1, input2, dim=1, eps=1e-08):
```
Comment (daemyung) [inline trident/function/function.py]: 함수를 설명하는 주석을 달아주세요.

File: trident/language/function.py [hunk 7]
```diff
@@ -60,6 +65,16 @@ def gemv(a, x):
     return triton.language.sum(a * triton.language.trans(x[:, None]), 1)
 
 
+@triton.jit
+def grad_sqrt(x):
```
Comment (daemyung) [inline trident/language/function.py]: 역전파 관련된 함수는 함수로 만들지 말아주세요.

File: trident/module.py [hunk 8]
```diff
@@ -173,6 +173,34 @@ def forward(self, input):
         return operation.Conv2d.apply(input, self.weight, self.bias)
 
 
+class CosineSimilarity(torch.nn.Module):
+    def __init__(self, dim: int = 1, eps: float = 1e-8):
```
Comment (daemyung) [inline trident/module.py:177]: 저희는 아규먼트의 타입을 정의하지 않습니다.

File: trident/module.py [hunk 9]
```diff
@@ -173,6 +173,34 @@ def forward(self, input):
...
+        """
+        Applies Cosine similarity to inputs.
```
Comment (daemyung) [inline trident/module.py]: Cosine을 cosine으로 변경해 주세요.

File: trident/module.py [hunk 10]
```diff
@@ -173,6 +173,34 @@ def forward(self, input):
...
+
+        Args:
+            dim: Dimension where cosine similarity is computed.
+            eps: Small value to avoid division by zero.
+        """
+        super().__init__()
+        self.dim = dim
+        self.eps = eps
+
+    def forward(self, input1, input2):
```
Comment (daemyung) [inline trident/module.py]: PyTorch 문서에는 변수명이 `x1`, `x2`로 되어 있긴 하네요.

File: trident/module.py [hunk 11]
```diff
@@ -173,6 +173,34 @@ def forward(self, input):
...
+
+        Args:
+             input1: an input1. Tensors of up to 3 dimensions
+             input2: an input2. Tensors of up to 3 dimensions.
+
+        Returns:
+            Cosine similarity between input1 and input2 along inserted dimension.
+        """
+
+        return operation.CosineSimilarity.apply(input1, input2, self.dim, self.eps)
```
Comment (daemyung) [inline trident/module.py]: `operation`에서 호출하지 말고 `function`을 호출해 주세요.

**Review by daemyung — CHANGES_REQUESTED** [review]

File: trident/module.py (hunk 8 above)
Comment (kakao-steve-ai) [reply trident/module.py:177]: 코딩스타일 관련 얘기인거 같은데, 나중에 한번 회의 해보면 좋을 것 같아요. 저는 개인적으로 아규먼트 타입 넣어주는게 버그도 줄여주고 좋은것 같아서요.

File: benchmarks/benchmark_cosine_similarity.py (hunk 3 above)
Comment (mejai1206) [reply benchmarks/benchmark_cosine_similarity.py]: `cosine similarity forward` 이렇게 변경하겠습니다.

File: benchmarks/benchmark_cosine_similarity.py (hunk 4 above)
Comment (mejai1206) [reply benchmarks/benchmark_cosine_similarity.py]: 구현상 z가 가장 외부의 차원인데, 배치일지는 어떻게 사용할지에 달린 것 같습니다.
torch 구현에선 차원의 제한이 따로 없더군요.
`sum` 과 비슷하다고 생각하여 `sum`을 참고하여 z, y, x 로 이름 지었습니다.

File: benchmarks/benchmark_cosine_similarity.py (hunk 1 above)
Comment (mejai1206) [reply benchmarks/benchmark_cosine_similarity.py]: 변경하겠습니다.

File: trident/module.py (hunk 11 above)
Comment (mejai1206) [reply trident/module.py]: 변경하겠습니다.

File: trident/module.py (hunk 10 above)
Comment (mejai1206) [reply trident/module.py]: `x1`, `x2` 변경하겠습니다.

File: benchmarks/benchmark_cosine_similarity.py (hunk 4 above)
Comment (daemyung) [reply benchmarks/benchmark_cosine_similarity.py]: 흠 코드를 작성하다보니 배치를 축으로 나타내면 코드를 작성하다가 헷갈리더라고요. 채널까진 괜찮은데,,,

File: trident/module.py (hunk 8 above)
Comment (daemyung) [reply trident/module.py:177]: 내일 티타임 시간에 이야기 하시죠.

File: trident/language/function.py (hunk 7 above)
Comment (mejai1206) [reply trident/language/function.py]: `grad_sqrt `  `grad_reciprocal` 삭제하겠습니다.

File: benchmarks/benchmark_cosine_similarity.py (hunk 4 above)
Comment (mejai1206) [reply benchmarks/benchmark_cosine_similarity.py]: 음.. 그럼 가장 외부차원인 `z_size` 만 `num_batches`라 바꾸고, 나머지 `y_size`, `x_size` 는 그대로 쓰는게 좋을까요?

Comment (mejai1206) [reply benchmarks/benchmark_cosine_similarity.py]: 논의결과대로 `z_size` --> `num_batches` 변경하겠습니다.

**Review by daemyung — CHANGES_REQUESTED** [review]
`triton.language`를 `tl`로 바꿔주세요.

File: trident/kernel/cosine_similarity.py [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+
+
+class CosineSimilarity:
+    @staticmethod
+    @triton.jit
+    def forward(
+        output_ptr,
+        denominator_ptr,
+        numerator_ptr,
+        input1_ptr,
```
Comment (daemyung) [inline trident/kernel/cosine_similarity.py]: `x1_ptr`, `x2_ptr`로 바꿔주세요.

**Comment by mejai1206** [comment]
cosine_similarity, language/constant, language/function
이부분 `tl` 로 수정되었습니다.

File: trident/kernel/cosine_similarity.py [hunk 13]
```diff
@@ -0,0 +1,445 @@
...
+            )
+
+        denominator_accumulation1 = tl.zeros((block_size, 1, 1), tl.float32)
+        denominator_accumulation2 = tl.zeros((block_size, 1, 1), tl.float32)
+        numerator_accumulation = tl.zeros((block_size, 1, 1), tl.float32)
+
+        for _ in range(0, size_along_dim, block_size):
+            input1 = tl.load(
+                input1_block_ptr, boundary_check=(0,), padding_option="zero"
+            ).to(tl.float32)
```
Comment (daemyung) [inline trident/kernel/cosine_similarity.py]: `float32`로 형변환을 하지 않으면 계산 결과가 틀리나요?

File: trident/operation/cosine_similarity.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+from trident import kernel, util
+
+
+class CosineSimilarity(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, input1, input2, dim, eps):
+        return CosineSimilarity.__forward(ctx, input1, input2, dim, eps)
+
+    @staticmethod
+    def __forward(ctx, input1, input2, dim, eps):
```
Comment (daemyung) [inline trident/operation/cosine_similarity.py]: `x1`, `x2`로 바꿔주세요.

**Review by daemyung — CHANGES_REQUESTED** [review]

File: trident/kernel/cosine_similarity.py (hunk 12 above)
Comment (mejai1206) [reply trident/kernel/cosine_similarity.py]: 변경하겠습니다.

File: trident/kernel/cosine_similarity.py (hunk 13 above)
Comment (mejai1206) [reply trident/kernel/cosine_similarity.py]: 순전파 커널에선 `x1`, `x2` 로드시 `to(tl.float32)` 를 하지 않아도 테스트를 통과하는군요.
이부분에서의 형변환은 생략 하도록 하겠습니다.

**Comment by daemyung** [comment]
만든 PR이 반영되기 전에 다른 PR들이 반영되면 저는 다음 방법으로 히스토리를 정리합니다.

1. 현재 커밋의 ID를 복사합니다.
2. `git reset --hard HEAD~1`
3. `git pull origin main`
4. `git cherry-pick {복사한 커밋 ID}`
5. `git push origin {브랜치 이름} --force`

File: trident/kernel/cosine_similarity.py [hunk 15]
```diff
@@ -0,0 +1,437 @@
...
+                order=(1, 0),
+            )
+
+        denominator_accumulation1 = tl.zeros((block_size, 1, 1), tl.float32)
+        denominator_accumulation2 = tl.zeros((block_size, 1, 1), tl.float32)
+        numerator_accumulation = tl.zeros((block_size, 1, 1), tl.float32)
+
+        for _ in range(0, size_along_dim, block_size):
+            x1 = tl.load(x1_block_ptr, boundary_check=(0,), padding_option="zero")
+
```
Comment (daemyung) [inline trident/kernel/cosine_similarity.py]: 불필요한 줄바꿈으로 보입니다.

File: trident/kernel/cosine_similarity.py (hunk 13 above)
Comment (mejai1206) [reply trident/kernel/cosine_similarity.py]: 역전파의
```
            denominator = tl.load(denominator_block_ptr).to(tl.float32)
            numerator = tl.load(numerator_block_ptr).to(tl.float32)
            grad_output = tl.load(grad_output_block_ptr).to(tl.float32)
```       
여기도 형변환 빼도 통과해서 추가로 삭제했습니다.

**Comment by mejai1206** [comment]
위 방법은 main 최신을 기반으로 리베이스한 것과 히스토리 결과가 다른걸까요?

**Comment by daemyung** [comment]
> 위 방법은 main 최신을 기반으로 리베이스한 것과 히스토리 결과가 다른걸까요?

무슨 말인지 이해를 못했어요. 저 방법이 최신 main을 기반으로 리베이스하는 거에요.

File: trident/kernel/cosine_similarity.py [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+        j = pid % projected_x_size
+
+        if dim == language.dim[0]:
+            x1_block_ptr = tl.make_block_ptr(
+                x1_ptr,
+                shape=(num_batches, y_size, x_size),
+                strides=(y_size * x_size, x_size, 1),
+                offsets=(0, i, j),
+                block_shape=(block_size, 1, 1),
+                order=(2, 1, 0),
```
Comment (kakao-steve-ai) [inline trident/kernel/cosine_similarity.py:51]: order를 어떻게 지정하는지에 따라 속도차이가 있나요?

File: trident/kernel/cosine_similarity.py [hunk 17]
```diff
@@ -0,0 +1,436 @@
...
+        dim: tl.constexpr,
+        block_size: tl.constexpr,
+        dtype: tl.constexpr,
+    ):
+        pid = tl.program_id(0)
+        projected_x_size = y_size if dim == 2 else x_size
+        i = pid // projected_x_size
+        j = pid % projected_x_size
+
+        if dim == language.dim[0]:
```
Comment (kakao-steve-ai) [inline trident/kernel/cosine_similarity.py:44]: 여기 if, elif, else에서 겹치는 코드가 많은거 같은데, 혹시 `shape`, `strides`만 앞에서 따로 계산하는거는 어떤가요?

**Comment by mejai1206** [comment]
main 최신 기반으로 충돌제거 및 리베이스했습니다.

File: trident/kernel/cosine_similarity.py (hunk 16 above)
Comment (mejai1206) [reply trident/kernel/cosine_similarity.py:51]: a30기준 벤치마크상으론 크게 변하는건 없었습니다만,
차원의 순서에 대한 order를 지정하는건, 특정 하드웨어에서 더 나은 성능을 주는데 도움이 된다고 적혀 있긴 했었습니다.

(참고: https://triton-lang.org/main/getting-started/tutorials/08-experimental-block-pointer.html )

Note that the order argument is set to (1, 0), which means the second axis is the inner dimension in terms of storage, and the first axis is the outer dimension. This information may sound redundant, but it is necessary for some hardware backends to optimize for better performance.

File: trident/kernel/cosine_similarity.py (hunk 17 above)
Comment (mejai1206) [reply trident/kernel/cosine_similarity.py:44]: `shape` `strides` 를 먼저 결정하고
이후에 `block_pointer` 를 만들고 싶었으나, 그렇게하면 트리톤의 버그인지 컴파일 에러가 나고,
아직 더 나은 방법을 못찾아서 일단 이렇게 하였습니다.

Comment (daemyung) [reply trident/kernel/cosine_similarity.py:44]: @ansteve 컴파일러가 죽어요.

Comment (kakao-steve-ai) [reply trident/kernel/cosine_similarity.py:44]: 넵 알겠습니다.

**Review by kakao-steve-ai — APPROVED** [review]
LGTM

**Review by daemyung — APPROVED** [review]
LGTM

File: trident/kernel/cosine_similarity.py (hunk 16 above)
Comment (daemyung) [reply trident/kernel/cosine_similarity.py:51]: @ansteve `row-major`, `column-major` 정도로 이해하고 있습니다.
